File: multi-db/src/main/java/com/multidb/service/impl/ApiServiceImpl.java
package com.multidb.service.impl;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.multidb.enums.Gender;
import com.multidb.repository.employee.EmployeeRepository;
import com.multidb.repository.employee.SalariesRepository;
import com.multidb.repository.employee.dto.EmployeeInfo;
import com.multidb.repository.employee.entity.EmployeeEntity;
import com.multidb.repository.employee.entity.SalariesEntity;
import com.multidb.repository.sakila.RentalRepository;
import com.multidb.repository.sakila.dto.MovieRentalInfo;
import com.multidb.service.ApiService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.ui.ModelMap;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class ApiServiceImpl implements ApiService {
    @Autowired
    private RentalRepository rentalRepository;

    @Autowired
    private SalariesRepository salariesRepository;

    @Autowired
    private EmployeeRepository employeeRepository;

    @Autowired
    @Qualifier("customObjectMapper")
    private ObjectMapper mapper;

    @Override
    public List<MovieRentalInfo> getRentalMovies(ModelMap params) {
        return rentalRepository.getRentalMovies((String) params.get("title"));
    }

    @Override
    public List<EmployeeInfo> getEmployeeSalary(Integer amount) {
        return salariesRepository.getEmployeeSalary(amount);
    }

    @Override
    @Transactional(transactionManager = "employeeTransactionManager")
    public void saveEmployee() {
        EmployeeEntity entity = employeeRepository.findById(253406).get();
        entity.setGender(Gender.M);
        employeeRepository.save(entity);
        int a = 1/0;
    }

    @Override
    @Transactional
    public void saveSalary() {

    }

    @Override
    @Transactional
    public void testSave() {
        saveEmployee();
        saveSalary();
    }
    @Override
    public List<?> handleData() throws JsonProcessingException {
//        var salaryData = salariesRepository.findAll()
        List<SalariesEntity> salaryData = salariesRepository.getSalaryTest();

//        List<EmployeeEntity> employeeData = employeeRepository.findAll();

        List<EmployeeEntity> result = employeeRepository.getEmployeeTest();

//        var result = employeeData.take(2)

        Long start = System.currentTimeMillis();
        System.out.printf("start: $start ********************************");
        for (EmployeeEntity s : result) {
            List<SalariesEntity> list = salaryData.stream()
                    .filter(it -> it.getEmpNo() == s.getEmpNo())
                    .collect(Collectors.toList());
            System.out.printf(mapper.writeValueAsString(list));
//            s.setSalary();
        }

        Long end = System.currentTimeMillis();
        System.out.printf("""
                Total time: ${end - start}
        """);
        System.out.printf("end: $end ********************************");

        Map<Integer, List<SalariesEntity>> salaryMap = salaryData.stream()
                .collect(Collectors.groupingBy(SalariesEntity::getEmpNo));
        return result;
    }

    @Override
    public Page<EmployeeEntity> getProductList(EmployeeEntity dto) {
        Pageable pageable = PageRequest.of(dto.getCurrentPage(), dto.getPageSize());
        return employeeRepository.getProductList(dto, pageable);
    }

}

File: mysql/src/main/kotlin/com/mysql/service/impl/ApiServiceImpl.kt
package com.mysql.service.impl

import com.fasterxml.jackson.databind.ObjectMapper
import com.mysql.enums.Gender
import com.mysql.repository.EmployeeRepository
import com.mysql.repository.SalariesRepository
import com.mysql.repository.entity.EmployeeEntity
import com.mysql.service.ApiService
import com.mysql.service.EmployeeService
import com.mysql.service.SalaryService
import com.mysql.service.TitleService
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


@Service
class ApiServiceImpl : ApiService {
    @Autowired
    @Qualifier("customObjectMapper")
    lateinit var mapper: ObjectMapper

    private val logger: Logger = LoggerFactory.getLogger(this.javaClass)

    @Autowired
    lateinit var salaryService: SalaryService

    @Autowired
    lateinit var employeeService: EmployeeService

    @Autowired
    lateinit var titleService: TitleService

    @Autowired
    lateinit var salariesRepository: SalariesRepository

    @Autowired
    lateinit var employeeRepository: EmployeeRepository

    @Transactional
    override fun testSave() {
        employeeService.saveEmployee(Gender.M)
        titleService.saveTitle("aaa")
        salaryService.saveSalary(40000)
    }

    override fun handleData(): List<*> {
        val salaryList = salariesRepository.findAll()

        val employeeData = employeeRepository.findAll()
        val result = employeeData.take(10).toMutableList()
        result.add(EmployeeEntity(empNo = 9911))

        val start = System.currentTimeMillis()
        println("start: $start ********************************")

        val salaryMap: Map<Int, Int> = salaryList
            .filter { it.salary != null }
            .groupBy({ it.empNo }, { it.salary!! })
            .mapValues { it.value.sumOf { it } }

//        var salaryTree: TreeMap<Int, List<SalariesEntity>> = TreeMap(salaryMap)

        for (emp in result) {
            emp.salary = salaryMap[emp.empNo]
        }

        val end = System.currentTimeMillis()
        println("Total time: ${end - start}")
        println("end: $end ********************************")
        return result
    }
}